File: frontend/apps/ios/Pantopus/Features/Scheduling/Home/Calendar/HomeAgendaComponents.swift
//
//  HomeAgendaComponents.swift
//  Pantopus
//
//  Stream I10 — the design's day-grouped agenda model + row card, shared by
//  the Home Calendar (F1) and the Permission-Gated Scheduler (F15) so the
//  time-led row, category chip, booking-union badge, and assignee avatar
//  stack read identically on both. Lifted from `home-shell.jsx` (`EventRow`,
//  `DaySection`) + `home-calendar-frames.jsx`.
//

import SwiftUI

// MARK: - Model

/// One agenda row, projected from a `CalendarEventDTO` (+ the member lookup).
public struct HomeAgendaItem: Sendable, Hashable, Identifiable {
    public let id: String
    /// "6:30" / "All day".
    public let time: String
    /// "PM" / "" (empty for all-day).
    public let ampm: String
    public let title: String
    public let category: CalendarEventCategory
    public let location: String?
    public let members: [HomeMember]
    /// Booking-union row (`source == "booking"`). Render-only, deep-links to E2.
    public let isBooking: Bool
    /// "pending" | "confirmed" (booking rows only).
    public let bookingStatus: String?
    /// The originating booking id (booking rows only) — for the E2 deep-link.
    public let bookingId: String?
    /// The home-event id (normal rows) — for the F2 detail push.
    public let eventId: String?
}

/// A day section in the agenda, with a relative header ("Today · Mon Jun 16").
public struct HomeAgendaSection: Sendable, Hashable, Identifiable {
    public let id: String
    public let header: String
    public let items: [HomeAgendaItem]
}

// MARK: - Builder

/// Pure projection from events → day-grouped agenda sections. Static so it's
/// trivially unit-testable. Mirrors `HomeCalendarViewModel`'s bucketing but
/// produces the richer `HomeAgendaItem` (assignees + booking-union) the
/// bespoke design row needs.
public enum HomeAgendaBuilder {
    public static func sections(
        events: [CalendarEventDTO],
        members: [String: HomeMember],
        now: Date,
        calendar: Calendar,
        timeZone: TimeZone,
        selectedIsoDate: String?,
        onlyUserId: String? = nil
    ) -> [HomeAgendaSection] {
        var cal = calendar
        cal.timeZone = timeZone

        let parsed: [(date: Date, dto: CalendarEventDTO)] = events
            .compactMap { dto in
                guard let date = parseInstant(dto.startAt) else { return nil }
                return (date, dto)
            }
            .filter { entry in
                guard let only = onlyUserId else { return true }
                return (entry.dto.assignedTo ?? []).contains(only)
            }
            .sorted { $0.date < $1.date }

        let todayStart = cal.startOfDay(for: now)

        // Group by ISO day, keeping only today + future (or the selected day).
        var buckets: [String: [HomeAgendaItem]] = [:]
        var order: [String] = []
        for entry in parsed {
            let iso = isoDay(entry.date, calendar: cal)
            if let selected = selectedIsoDate {
                guard iso == selected else { continue }
            } else {
                guard entry.date >= todayStart else { continue }
            }
            let item = item(from: entry.dto, start: entry.date, members: members, calendar: cal)
            if buckets[iso] == nil { order.append(iso) }
            buckets[iso, default: []].append(item)
        }

        return order.map { iso in
            HomeAgendaSection(
                id: iso,
                header: header(forIso: iso, now: now, calendar: cal),
                items: buckets[iso] ?? []
            )
        }
    }

    // MARK: Item projection

    static func item(
        from dto: CalendarEventDTO,
        start: Date,
        members: [String: HomeMember],
        calendar cal: Calendar
    ) -> HomeAgendaItem {
        let allDay = dto.endAt == nil && isMidnight(start, calendar: cal)
        let (time, ampm) = timeParts(start, allDay: allDay, calendar: cal)
        let assignees = (dto.assignedTo ?? []).compactMap { members[$0] }
        let isBooking = dto.source == "booking"
        return HomeAgendaItem(
            id: dto.id,
            time: time,
            ampm: ampm,
            title: dto.title,
            category: CalendarEventCategory.from(eventType: dto.eventType),
            location: dto.locationNotes?.isEmpty == false ? dto.locationNotes : nil,
            members: assignees,
            isBooking: isBooking,
            bookingStatus: dto.bookingStatus,
            bookingId: dto.bookingId,
            eventId: isBooking ? nil : dto.id
        )
    }

    // MARK: Formatting

    static func timeParts(_ date: Date, allDay: Bool, calendar cal: Calendar) -> (String, String) {
        if allDay { return ("All day", "") }
        let timeFmt = DateFormatter()
        timeFmt.locale = Locale(identifier: "en_US_POSIX")
        timeFmt.timeZone = cal.timeZone
        timeFmt.dateFormat = "h:mm"
        let ampmFmt = DateFormatter()
        ampmFmt.locale = Locale(identifier: "en_US_POSIX")
        ampmFmt.timeZone = cal.timeZone
        ampmFmt.dateFormat = "a"
        return (timeFmt.string(from: date), ampmFmt.string(from: date))
    }

    static func header(forIso iso: String, now: Date, calendar cal: Calendar) -> String {
        guard let date = parseIso(iso, calendar: cal) else { return iso }
        let dayFmt = DateFormatter()
        dayFmt.locale = Locale(identifier: "en_US_POSIX")
        dayFmt.timeZone = cal.timeZone
        dayFmt.dateFormat = "EEE MMM d"
        let stamp = dayFmt.string(from: date)
        let todayStart = cal.startOfDay(for: now)
        let dayStart = cal.startOfDay(for: date)
        let delta = cal.dateComponents([.day], from: todayStart, to: dayStart).day ?? 0
        switch delta {
        case 0: return "Today · \(stamp)"
        case 1: return "Tomorrow · \(stamp)"
        default: return stamp
        }
    }

    static func isMidnight(_ date: Date, calendar cal: Calendar) -> Bool {
        let parts = cal.dateComponents([.hour, .minute, .second], from: date)
        return (parts.hour ?? 0) == 0 && (parts.minute ?? 0) == 0 && (parts.second ?? 0) == 0
    }

    static func isoDay(_ date: Date, calendar cal: Calendar) -> String {
        let fmt = DateFormatter()
        fmt.locale = Locale(identifier: "en_US_POSIX")
        fmt.timeZone = cal.timeZone
        fmt.dateFormat = "yyyy-MM-dd"
        return fmt.string(from: date)
    }

    static func parseIso(_ iso: String, calendar cal: Calendar) -> Date? {
        let fmt = DateFormatter()
        fmt.locale = Locale(identifier: "en_US_POSIX")
        fmt.timeZone = cal.timeZone
        fmt.dateFormat = "yyyy-MM-dd"
        return fmt.date(from: iso)
    }

    static func parseInstant(_ iso: String) -> Date? {
        let withFrac = ISO8601DateFormatter()
        withFrac.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let d = withFrac.date(from: iso) { return d }
        let plain = ISO8601DateFormatter()
        plain.formatOptions = [.withInternetDateTime]
        if let d = plain.date(from: iso) { return d }
        let dayFmt = DateFormatter()
        dayFmt.locale = Locale(identifier: "en_US_POSIX")
        dayFmt.timeZone = TimeZone(identifier: "UTC")
        dayFmt.dateFormat = "yyyy-MM-dd"
        return dayFmt.date(from: iso)
    }

    // MARK: Month strip

    /// First day of the week (Sunday) containing `date`, ISO yyyy-MM-dd.
    static func weekAnchorIso(for date: Date, calendar cal: Calendar) -> String {
        let weekday = cal.component(.weekday, from: date) // 1 = Sunday
        let daysBack = weekday - cal.firstWeekday
        let anchor = cal.date(byAdding: .day, value: -daysBack, to: cal.startOfDay(for: date))
            ?? cal.startOfDay(for: date)
        return isoDay(anchor, calendar: cal)
    }

    /// Build a 7-day `MonthStripState` for the week anchored at `anchorIso`,
    /// counting events per day. Used by the gated scheduler (F15).
    static func weekStrip(
        events: [CalendarEventDTO],
        anchorIso: String,
        selectedIso: String?,
        now: Date,
        calendar: Calendar,
        timeZone: TimeZone
    ) -> MonthStripState? {
        var cal = calendar
        cal.timeZone = timeZone
        guard let anchor = parseIso(anchorIso, calendar: cal) else { return nil }

        var dotCounts: [String: Int] = [:]
        for dto in events {
            guard let date = parseInstant(dto.startAt) else { continue }
            dotCounts[isoDay(date, calendar: cal), default: 0] += 1
        }

        let monthFmt = DateFormatter()
        monthFmt.locale = Locale(identifier: "en_US_POSIX")
        monthFmt.timeZone = cal.timeZone
        monthFmt.dateFormat = "MMMM yyyy"
        let dowFmt = DateFormatter()
        dowFmt.locale = Locale(identifier: "en_US_POSIX")
        dowFmt.timeZone = cal.timeZone
        dowFmt.dateFormat = "EEE"
        let dayFmt = DateFormatter()
        dayFmt.locale = Locale(identifier: "en_US_POSIX")
        dayFmt.timeZone = cal.timeZone
        dayFmt.dateFormat = "d"

        var days: [MonthStripState.Day] = []
        for offset in 0..<7 {
            guard let date = cal.date(byAdding: .day, value: offset, to: anchor) else { continue }
            let iso = isoDay(date, calendar: cal)
            days.append(
                MonthStripState.Day(
                    id: iso,
                    dayOfWeek: dowFmt.string(from: date),
                    date: Int(dayFmt.string(from: date)) ?? 0,
                    eventCount: dotCounts[iso] ?? 0
                )
            )
        }
        return MonthStripState(
            monthLabel: monthFmt.string(from: anchor),
            days: days,
            selectedIsoDate: selectedIso,
            todayIsoDate: isoDay(now, calendar: cal)
        )
    }
}

// MARK: - Row card

/// The design's `EventRow`: a 42pt time column, a 1px divider, the title with
/// a category chip + booking badge + location, and a trailing assignee stack.
public struct HomeAgendaRowCard: View {
    let item: HomeAgendaItem
    var dimmed: Bool = false
    let onTap: @MainActor () -> Void

    public init(item: HomeAgendaItem, dimmed: Bool = false, onTap: @escaping @MainActor () -> Void) {
        self.item = item
        self.dimmed = dimmed
        self.onTap = onTap
    }

    public var body: some View {
        Button(action: onTap) {
            HStack(spacing: Spacing.s3) {
                VStack(spacing: 1) {
                    Text(item.time)
                        .font(.system(size: 13, weight: .bold))
                        .foregroundStyle(Theme.Color.appText)
                        .monospacedDigit()
                    if !item.ampm.isEmpty {
                        Text(item.ampm)
                            .font(.system(size: 9.5, weight: .semibold))
                            .foregroundStyle(Theme.Color.appTextMuted)
                    }
                }
                .frame(width: 42)

                Rectangle().fill(Theme.Color.appBorder).frame(width: 1)

                VStack(alignment: .leading, spacing: 4) {
                    Text(item.title)
                        .font(.system(size: 13.5, weight: .bold))
                        .foregroundStyle(Theme.Color.appText)
                        .lineLimit(1)
                    HStack(spacing: Spacing.s1) {
                        CategoryChipMini(category: item.category)
                        if item.isBooking {
                            HomeBookingTag()
                            if let status = item.bookingStatus {
                                SchedulingStatusPill(status: status)
                            }
                        }
                        if let location = item.location {
                            HStack(spacing: 3) {
                                Icon(.mapPin, size: 10, color: Theme.Color.appTextSecondary)
                                Text(location)
                                    .font(.system(size: 10.5))
                                    .foregroundStyle(Theme.Color.appTextSecondary)
                                    .lineLimit(1)
                            }
                        }
                    }
                }
                Spacer(minLength: Spacing.s1)
                if !item.members.isEmpty {
                    HomeAvatarStack(members: item.members, size: 26)
                }
            }
            .padding(.horizontal, Spacing.s3)
            .padding(.vertical, 11)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Theme.Color.appSurface)
            .clipShape(RoundedRectangle(cornerRadius: Radii.xl, style: .continuous))
            .overlay(
                RoundedRectangle(cornerRadius: Radii.xl, style: .continuous)
                    .stroke(Theme.Color.appBorder, lineWidth: 1)
            )
            .opacity(dimmed ? 0.55 : 1)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityElement(children: .combine)
        .accessibilityLabel(accessibilityLabel)
        .accessibilityIdentifier("homeAgendaRow_\(item.id)")
    }

    private var accessibilityLabel: String {
        var parts = ["\(item.time) \(item.ampm)", item.title, item.category.label]
        if item.isBooking { parts.append("Booking") }
        if let location = item.location { parts.append(location) }
        return parts.joined(separator: ", ")
    }
}

/// The design's small category chip (sunken pill + colour dot + label).
public struct CategoryChipMini: View {
    let category: CalendarEventCategory

    public init(category: CalendarEventCategory) {
        self.category = category
    }

    public var body: some View {
        HStack(spacing: Spacing.s1) {
            Circle().fill(category.foreground).frame(width: 7, height: 7)
            Text(category.label)
                .font(.system(size: 10, weight: .semibold))
                .foregroundStyle(Theme.Color.appTextStrong)
        }
        .padding(.horizontal, 7)
        .padding(.vertical, 2)
        .background(Theme.Color.appSurfaceSunken)
        .clipShape(Capsule())
    }
}

/// Shimmer skeleton row mirroring the loaded geometry (never a spinner).
public struct HomeAgendaSkeletonRow: View {
    public init() {}
    public var body: some View {
        HStack(spacing: Spacing.s3) {
            VStack(spacing: 4) {
                Shimmer(width: 30, height: 11)
                Shimmer(width: 20, height: 8)
            }
            .frame(width: 42)
            Rectangle().fill(Theme.Color.appBorder).frame(width: 1, height: 36)
            VStack(alignment: .leading, spacing: 6) {
                Shimmer(width: 150, height: 11)
                Shimmer(width: 90, height: 9)
            }
            Spacer()
            Shimmer(width: 26, height: 26, cornerRadius: 13)
        }
        .padding(.horizontal, Spacing.s3)
        .padding(.vertical, 11)
        .background(Theme.Color.appSurface)
        .clipShape(RoundedRectangle(cornerRadius: Radii.xl, style: .continuous))
        .overlay(
            RoundedRectangle(cornerRadius: Radii.xl, style: .continuous)
                .stroke(Theme.Color.appBorder, lineWidth: 1)
        )
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/homes/calendar/HomeAgendaComponents.kt
@file:Suppress("PackageNaming", "MagicNumber", "TooManyFunctions", "LongMethod", "LongParameterList")

package app.pantopus.android.ui.screens.homes.calendar

import androidx.compose.foundation.ScrollState
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.pantopus.android.data.api.models.homes.CalendarEventDto
import app.pantopus.android.ui.components.Shimmer
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusTextStyle
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.util.Locale

// ─── Projection models ────────────────────────────────────────────────────

/** One agenda row. `isBooking` rows are read-only (eventId is null). */
@Immutable
data class HomeAgendaItem(
    val id: String,
    val time: String,
    val ampm: String,
    val title: String,
    val category: CalendarEventCategory,
    val location: String?,
    val members: List<HomeMember>,
    val isBooking: Boolean,
    val bookingStatus: String?,
    val bookingId: String?,
    val eventId: String?,
)

/** A day-grouped agenda section. */
@Immutable
data class HomeAgendaSection(
    val id: String,
    val header: String,
    val items: List<HomeAgendaItem>,
)

/** The member-scoped agenda filter. */
sealed interface MemberFilter {
    data object All : MemberFilter

    data object Mine : MemberFilter

    data class Member(val id: String, val name: String) : MemberFilter
}

/** Empty-agenda reasons. */
sealed interface AgendaEmpty {
    data object FirstRun : AgendaEmpty

    data class FilteredMember(val name: String) : AgendaEmpty

    data object FilteredDay : AgendaEmpty
}

// ─── Pure projection ──────────────────────────────────────────────────────

/**
 * Pure agenda/month-strip projection. Mirrors iOS `HomeAgendaBuilder` —
 * UTC-anchored, Sunday-first, deterministic for unit tests.
 */
object HomeAgendaBuilder {
    private val isoDate = DateTimeFormatter.ISO_DATE
    private val headerFmt = DateTimeFormatter.ofPattern("EEE MMM d", Locale.US)
    private val timeFmt = DateTimeFormatter.ofPattern("h:mm", Locale.US)
    private val ampmFmt = DateTimeFormatter.ofPattern("a", Locale.US)
    private val monthFmt = DateTimeFormatter.ofPattern("MMMM yyyy", Locale.US)
    private val dowFmt = DateTimeFormatter.ofPattern("EEE", Locale.US)

    fun parseInstant(iso: String?): Instant? {
        if (iso.isNullOrBlank()) return null
        return runCatching { Instant.parse(iso) }
            .recoverCatching { LocalDate.parse(iso).atStartOfDay(ZoneId.of("UTC")).toInstant() }
            .getOrNull()
    }

    private fun isoDay(
        instant: Instant,
        zone: ZoneId,
    ): String = instant.atZone(zone).toLocalDate().format(isoDate)

    private fun isMidnight(zoned: ZonedDateTime): Boolean = zoned.hour == 0 && zoned.minute == 0 && zoned.second == 0

    /** Project a single DTO into an agenda item. */
    fun item(
        dto: CalendarEventDto,
        start: Instant,
        members: Map<String, HomeMember>,
        zone: ZoneId,
    ): HomeAgendaItem {
        val zoned = start.atZone(zone)
        val allDay = dto.endAt == null && isMidnight(zoned)
        val time = if (allDay) "All day" else timeFmt.format(zoned)
        val ampm = if (allDay) "" else ampmFmt.format(zoned)
        val assigned = dto.assignedTo.orEmpty().mapNotNull { members[it] }
        return HomeAgendaItem(
            id = dto.id,
            time = time,
            ampm = ampm,
            title = dto.title,
            category = CalendarEventCategory.from(dto.eventType),
            location = dto.locationNotes?.takeIf { it.isNotBlank() },
            members = assigned,
            isBooking = dto.isBooking,
            bookingStatus = dto.bookingStatus,
            bookingId = dto.bookingId,
            eventId = if (dto.isBooking) null else dto.id,
        )
    }

    /** Day-grouped sections. `selectedIsoDate` pins one day; else past events drop. */
    fun sections(
        events: List<CalendarEventDto>,
        members: Map<String, HomeMember>,
        now: Instant,
        zone: ZoneId,
        selectedIsoDate: String? = null,
        onlyUserId: String? = null,
    ): List<HomeAgendaSection> {
        val parsed =
            events
                .mapNotNull { dto -> parseInstant(dto.startAt)?.let { dto to it } }
                .filter { (dto, _) -> onlyUserId == null || dto.assignedTo.orEmpty().contains(onlyUserId) }
                .sortedBy { it.second }
        val todayStart = now.atZone(zone).toLocalDate().atStartOfDay(zone).toInstant()
        val kept =
            if (selectedIsoDate != null) {
                parsed.filter { isoDay(it.second, zone) == selectedIsoDate }
            } else {
                parsed.filter { !it.second.isBefore(todayStart) }
            }
        return kept
            .groupBy { isoDay(it.second, zone) }
            .toSortedMap()
            .map { (iso, bucket) ->
                HomeAgendaSection(
                    id = iso,
                    header = header(iso, now, zone),
                    items = bucket.map { (dto, start) -> item(dto, start, members, zone) },
                )
            }
    }

    private fun header(
        iso: String,
        now: Instant,
        zone: ZoneId,
    ): String {
        val date = runCatching { LocalDate.parse(iso) }.getOrNull() ?: return iso
        val stamp = headerFmt.format(date)
        val today = now.atZone(zone).toLocalDate()
        return when (java.time.temporal.ChronoUnit.DAYS.between(today, date)) {
            0L -> "Today · $stamp"
            1L -> "Tomorrow · $stamp"
            else -> stamp
        }
    }

    /** Sunday-anchored start-of-week ISO date for [now]. */
    fun weekAnchorIso(
        now: Instant,
        zone: ZoneId,
    ): String {
        val date = now.atZone(zone).toLocalDate()
        val back =
            when (date.dayOfWeek) {
                java.time.DayOfWeek.SUNDAY -> 0
                java.time.DayOfWeek.MONDAY -> 1
                java.time.DayOfWeek.TUESDAY -> 2
                java.time.DayOfWeek.WEDNESDAY -> 3
                java.time.DayOfWeek.THURSDAY -> 4
                java.time.DayOfWeek.FRIDAY -> 5
                java.time.DayOfWeek.SATURDAY -> 6
            }
        return date.minusDays(back.toLong()).format(isoDate)
    }

    /** Build the 7-day month-strip state from the visible week anchor. */
    fun weekStrip(
        events: List<CalendarEventDto>,
        anchorIso: String,
        selectedIso: String?,
        now: Instant,
        zone: ZoneId,
    ): MonthStripState {
        val anchor = runCatching { LocalDate.parse(anchorIso) }.getOrNull() ?: now.atZone(zone).toLocalDate()
        val dotCounts = mutableMapOf<String, Int>()
        for (dto in events) {
            val start = parseInstant(dto.startAt) ?: continue
            val iso = isoDay(start, zone)
            dotCounts[iso] = (dotCounts[iso] ?: 0) + 1
        }
        val days =
            (0 until 7).map { offset ->
                val date = anchor.plusDays(offset.toLong())
                val iso = date.format(isoDate)
                MonthStripState.Day(
                    id = iso,
                    dayOfWeek = dowFmt.format(date),
                    date = date.dayOfMonth,
                    eventCount = dotCounts[iso] ?: 0,
                )
            }
        return MonthStripState(
            monthLabel = monthFmt.format(anchor),
            days = days,
            selectedIsoDate = selectedIso,
            todayIsoDate = isoDay(now, zone),
        )
    }
}

// ─── Row components ───────────────────────────────────────────────────────

/** The agenda union row. Mirrors iOS `HomeAgendaRowCard`. */
@Composable
fun HomeAgendaRowCard(
    item: HomeAgendaItem,
    modifier: Modifier = Modifier,
    dimmed: Boolean = false,
    enabled: Boolean = true,
    onTap: () -> Unit = {},
) {
    val rowLabel =
        buildString {
            append("${item.time} ${item.ampm}, ${item.title}, ${item.category.label}")
            if (item.isBooking) append(", Booking")
            item.location?.let { append(", $it") }
        }
    Row(
        modifier =
            modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.xl))
                .background(PantopusColors.appSurface)
                .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.xl))
                .then(if (enabled) Modifier.clickable(onClick = onTap) else Modifier)
                .padding(horizontal = Spacing.s3, vertical = 11.dp)
                .alpha(if (dimmed) 0.55f else 1f)
                .testTag("homeAgendaRow_${item.id}")
                .semantics { contentDescription = rowLabel },
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.s3),
    ) {
        Column(
            modifier = Modifier.width(42.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(1.dp),
        ) {
            Text(text = item.time, fontSize = 13.sp, fontWeight = FontWeight.Bold, color = PantopusColors.appText)
            if (item.ampm.isNotEmpty()) {
                Text(text = item.ampm, fontSize = 9.5.sp, fontWeight = FontWeight.SemiBold, color = PantopusColors.appTextMuted)
            }
        }
        Box(modifier = Modifier.width(1.dp).height(36.dp).background(PantopusColors.appBorder))
        Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(4.dp)) {
            Text(
                text = item.title,
                fontSize = 13.5.sp,
                fontWeight = FontWeight.Bold,
                color = PantopusColors.appText,
                maxLines = 1,
            )
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(Spacing.s1),
            ) {
                CategoryChipMini(category = item.category)
                if (item.isBooking) {
                    HomeBookingTag()
                    item.bookingStatus?.let { SchedulingStatusBadge(status = it) }
                }
                item.location?.let { loc ->
                    Text(text = loc, fontSize = 10.5.sp, color = PantopusColors.appTextSecondary, maxLines = 1)
                }
            }
        }
        if (item.members.isNotEmpty()) {
            HomeAvatarStack(members = item.members, size = 26.dp)
        }
    }
}

/** Mini category chip — dot + label. Mirrors iOS `CategoryChipMini`. */
@Composable
fun CategoryChipMini(
    category: CalendarEventCategory,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier =
            modifier
                .clip(RoundedCornerShape(percent = 50))
                .background(PantopusColors.appSurfaceSunken)
                .padding(horizontal = 7.dp, vertical = 2.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.s1),
    ) {
        Box(modifier = Modifier.size(7.dp).clip(CircleShape).background(category.foreground))
        Text(text = category.label, fontSize = 10.sp, fontWeight = FontWeight.SemiBold, color = PantopusColors.appTextStrong)
    }
}

/** Loading skeleton row mirroring the loaded geometry. */
@Composable
fun HomeAgendaSkeletonRow(modifier: Modifier = Modifier) {
    Row(
        modifier =
            modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.xl))
                .background(PantopusColors.appSurface)
                .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.xl))
                .padding(horizontal = Spacing.s3, vertical = 11.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.s3),
    ) {
        Column(
            modifier = Modifier.width(42.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(4.dp),
        ) {
            Shimmer(width = 30.dp, height = 11.dp, cornerRadius = Radii.xs)
            Shimmer(width = 20.dp, height = 8.dp, cornerRadius = Radii.xs)
        }
        Box(modifier = Modifier.width(1.dp).height(36.dp).background(PantopusColors.appBorder))
        Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(6.dp)) {
            Shimmer(width = 150.dp, height = 11.dp, cornerRadius = Radii.xs)
            Shimmer(width = 90.dp, height = 9.dp, cornerRadius = Radii.xs)
        }
        Shimmer(width = 26.dp, height = 26.dp, cornerRadius = Radii.pill)
    }
}

/** Horizontal member-filter chip row. Mirrors iOS `FilterChipRow`. */
@Composable
fun FilterChipRow(
    chips: List<MemberFilter>,
    selected: MemberFilter,
    onSelect: (MemberFilter) -> Unit,
    modifier: Modifier = Modifier,
    scrollState: ScrollState = rememberScrollState(),
) {
    Row(
        modifier =
            modifier
                .fillMaxWidth()
                .background(PantopusColors.appSurface)
                .horizontalScroll(scrollState)
                .padding(horizontal = Spacing.s3, vertical = Spacing.s2),
        horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        chips.forEach { chip ->
            val label = filterLabel(chip)
            val active = chip == selected
            Row(
                modifier =
                    Modifier
                        .heightIn(min = 30.dp)
                        .clip(RoundedCornerShape(percent = 50))
                        .background(if (active) PantopusColors.homeBg else PantopusColors.appSurface)
                        .then(
                            if (active) {
                                Modifier
                            } else {
                                Modifier.border(1.dp, PantopusColors.appBorder, RoundedCornerShape(percent = 50))
                            },
                        ).clickable { onSelect(chip) }
                        .padding(horizontal = 13.dp, vertical = 6.dp)
                        .testTag("homeCalendar_filter_$label")
                        .semantics { selected = active },
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(
                    text = label,
                    fontSize = 12.sp,
                    fontWeight = if (active) FontWeight.Bold else FontWeight.SemiBold,
                    color = if (active) PantopusColors.homeDark else PantopusColors.appTextStrong,
                )
            }
        }
    }
}

/** Filter chip label: All / Mine / first word of a member's name. */
fun filterLabel(filter: MemberFilter): String =
    when (filter) {
        MemberFilter.All -> "All"
        MemberFilter.Mine -> "Mine"
        is MemberFilter.Member -> filter.name.split(' ').firstOrNull().orEmpty().ifEmpty { filter.name }
    }

/** Section header text for the bespoke agenda body. */
@Composable
fun AgendaSectionHeader(
    text: String,
    modifier: Modifier = Modifier,
) {
    Text(
        text = text,
        style = PantopusTextStyle.caption,
        fontWeight = FontWeight.Bold,
        color = PantopusColors.appTextSecondary,
        modifier = modifier.padding(horizontal = Spacing.s1, vertical = Spacing.s1),
    )
}
